### PR TITLE
[ATen][FakeTensor] Handle unbacked dims in folded matmul

### LINKED ISSUE #3322: [graph_trainer] dynamic chunking blocked by rank-3 biasless linear with mark_unbacked
# GraphTrainer dynamic chunking: `mark_unbacked` batch dim fails on biasless rank-3 `linear`

## Summary

GraphTrainer chunk passes need to mark selected activation dimensions dynamic before tracing so the chunked graph can split batch/sequence dimensions symbolically. This currently works for MoE-only regions, but tracing an attention-containing TransformerBlock with dynamic batch fails before the chunk pass runs.

The failure is caused by biasless rank-3 `nn.Linear` / `F.linear` on an input with an unbacked symbolic leading dimension. This is hit by DeepSeekV3 attention projections such as `self.wq(x)`, where `x` is `[batch, seq, hidden]` and the projection has `bias=False`.

## Minimal Repro

Run from a TorchTitan checkout:

```bash
 python /tmp/repro_graph_trainer_unbacked_rank3_linear.py
```

Script:

```python
import traceback

import torch
import torch.nn.functional as F
from torch._dynamo.decorators import mark_unbacked

from torchtitan.experiments.graph_trainer.make_fx_tracer import (
    minimal_fx_tracer,
    run_traced,
)


def trace_case(name, fn, x_shape):
    print(f"\n{name}")
    x = torch.randn(*x_shape, device="cuda")
    w = torch.randn(3072, 256, device="cuda")
    mark_unbacked(x, 0, min=2, max=8, hint_override=x_shape[0])

    try:
        traced = minimal_fx_tracer(lambda state, x: fn(x, state["w"]))(
            {"w": w},
            x,
        )
        replay_x = torch.randn(5, *x_shape[1:], device="cuda")
        y = run_traced(traced, {"w": w}, replay_x)
        print("OK", tuple(y.shape))
    except Exception as exc:
        print(type(exc).__name__)
        print(exc)
        traceback.print_exc(limit=8)


def linear_2d(x, w):
    return F.linear(x, w)


def linear_3d_biasless(x, w):
    return F.linear(x, w)


def linear_3d_explicit_flatten(x, w):
    bsz = x.size(0)
    seqlen = x.size(1)
    y = F.linear(x.reshape(-1, x.size(-1)), w)
    return y.reshape(bsz, seqlen, -1)


if __name__ == "__main__":
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for this repro")

    trace_case("2D biasless linear: expected OK", linear_2d, (4, 256))
    trace_case("3D biasless linear: currently fails", linear_3d_biasless, (4, 16, 256))
    trace_case(
        "3D explicit flatten + linear + reshape: expected OK",
        linear_3d_explicit_flatten,
        (4, 16, 256),
    )
```

Observed output:

```text
2D biasless linear: expected OK
OK (5, 3072)

3D biasless linear: currently fails
RuntimeError
Cannot call numel() on tensor with symbolic sizes/strides

3D explicit flatten + linear + reshape: expected OK
OK (5, 16, 3072)
```

## Why This Matters

GraphTrainer chunk passes need to trace modules with symbolic batch/sequence dimensions. For DSv3 MoE-only regions, this works and we can validate bitwise equivalence against eager chunking. For a full TransformerBlock, tracing fails at the first attention projection before chunking has a chance to run.

The relevant model path is:

```text
DeepSeekV3TransformerBlock.forward
-> Attention.forward
-> q = self.wq(x)
-> biasless rank-3 F.linear(x, weight)
-> GuardOnDataDependentSymNode
```

## Root Cause Hypothesis

2D biasless linear lowers to `mm`, whose meta implementation can preserve the symbolic leading dimension.

3D biasless linear goes through the `matmul` path. The matmul folding decision checks emptiness / folding conditions using a `numel()`-derived expression. With input shape `[u0, 16, 256]`, that expression depends on unbacked symbol `u0`. Even with positive bounds from `mark_unbacked(x, 0, min=2, max=8)`, the current fake/meta path asks for a concrete/specialized integer and raises.

In the full DeepSeekV3 TransformerBlock trace this may surface as either:

```text
Cannot call numel() on tensor with symbolic sizes/strides
```

or:

```text
GuardOnDataDependentSymNode: Could not extract specialized integer from data-dependent expression u0
```

depending on the exact tracing path.

## Desired Behavior

Tracing should succeed without specializing the batch dimension to the hint. The resulting graph should keep the leading dimension symbolic and replay with another valid batch size.

## Upstream PyTorch Fix Direction

Preferred upstream direction:

> Teach the `matmul` folding emptiness check to reason from symbolic bounds without forcing a concrete integer.

Specifically, if an unbacked symbol has a positive lower bound, `numel() == 0` for shapes like `[u0, 16, 256]` should be provably false, so the folding decision should not require concretizing `u0`.

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stacked PRs:
 * #183545
 * #183544
 * #183495
 * #183398
 * __->__#183397

--- --- ---

### [ATen][FakeTensor] Handle unbacked dims in folded matmul

The folded matmul path used for ND @ 1D/2D tensors was still forcing concrete sizes in two places: the empty-tensor check used numel(), and the fold/view path built concrete DimVector shapes from sizes(). With unbacked input dimensions, both paths can require specializing a data-dependent symbol even though the operation can be represented symbolically.

Keep the folded path symbolic by using sym_numel(), sym_sizes(), sym_strides(), reshape_symint(), _unsafe_view_symint(), and symbolic output resize. The contiguity predicate now uses symbolic equality guarded through TORCH_GUARD_OR_FALSE instead of forcing concrete stride/shape integers.

Also let fake view metadata recover from hinted unbacked contiguity relations by emitting symbolic equality checks for stride relations that are true at the traced hint. This handles shapes such as 2 * (u0 // 2) with a stride expressed through u0 without specializing u0 to a concrete value.

Add fake tensor regression tests for rank-3 linear with an unbacked batch dimension and hinted unbacked view contiguity.

Test Plan:

python test/test_fake_tensor.py FakeTensorTest.test_linear_rank3_unbacked_batch_dim FakeTensorTest.test_view_unbacked_contiguous_relation_from_hints -v

lintrunner --config=.lintrunner.toml aten/src/ATen/native/LinearAlgebra.cpp torch/_decomp/decompositions.py torch/_subclasses/fake_impls.py test/test_fake_tensor.py

Fixes pytorch/torchtitan#3322

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -1979,16 +1981,17 @@ static bool should_fold(const Tensor& tensor1, const Tensor& tensor2, bool has_o
...
 
   // t1->view(-1, t1->size(-1)) does not copy only when the first n-1 dimensions are contiguous
   // in the sense that t1_stride[i] = t1_stride[i+1]*t1_shape[i+1]
-  const auto t1_shape = t1->sizes();
-  const auto t1_strides = t1->strides();
+  const auto t1_shape = t1->sym_sizes();
+  const auto t1_strides = t1->sym_strides();
   for (auto i = int64_t{0}; i < dim_t1 - int64_t{2}; ++i) {
-    if (t1_strides[i] != t1_strides[i+1] * t1_shape[i+1]) {
+    if (!TORCH_GUARD_OR_FALSE(
```
Comment (laithsakka) [inline aten/src/ATen/native/LinearAlgebra.cpp]: nit:
a) 
```
if (TORCH_GUARD_OR_TRUE(t1_strides[i] != t1_strides[i+1] * t1_shape[i+1])) {
  return false;
}
```
is easier to follow on

**Review by laithsakka — APPROVED** [review]

File: test/test_fake_tensor.py [hunk 2]
```diff
@@ -902,6 +892,29 @@ def test_same_shape_env_preserved(self):
         self.assertIs(t2.size(0).node.shape_env, t1.size(0).node.shape_env)
         self.assertEqual(str(t2.size(0)), str(t1.size(0)))
 
+    def test_linear_rank3_unbacked_batch_dim(self):
```
Comment (laithsakka) [inline test/test_fake_tensor.py:906]: make sure this repo the original issue i did not verify

Comment (sanketpurandare) [reply test/test_fake_tensor.py:906]: The regression now matches the original pytorch/torchtitan#3322 failure mode
more directly: it still checks the fake/meta output keeps the leading dimension
unbacked, and it now also traces biasless rank-3 `F.linear` with
`mark_unbacked` + symbolic `make_fx` and replays the graph at a different batch
size.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (sanketpurandare) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Updated this to guard the symbolic inequality directly with `TORCH_GUARD_OR_TRUE(...sym_ne(...))`, which makes the negative condition clearer and matches the existing symbolic predicate style in ATen.
